myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md

```diff
@@ -0,0 +1,747 @@
+import arviz as az
+import numpy as np
+import pandas as pd
+import pymc as pm
+
+from matplotlib import pyplot as plt
+```
+
+```{code-cell} ipython3
+RANDOM_SEED = 8929
+rng = np.random.default_rng(RANDOM_SEED)
+az.style.use("arviz-darkgrid")
+```
+
+## Generate Fake Autoregressive Data
+
+First we will generate a simple autoregressive timeseries. We will show how to specify a model to fit this data and then add a number of complexities to the data and show how they too can be captured with an autoregressive model and used to predict the shape of the future.
+
+```{code-cell} ipython3
+def simulate_ar(intercept, coef1, coef2, noise=0.3, *, warmup=10, steps=200):
+    # We sample some extra warmup steps, to let the AR process stabilize
+    draws = np.zeros(warmup + steps)
+    # Initialize first draws at intercept
+    draws[:2] = intercept
+    for step in range(2, warmup + steps):
+        draws[step] = (
+            intercept
+            + coef1 * draws[step - 1]
+            + coef2 * draws[step - 2]
+            + np.random.normal(0, noise)
+        )
+    # Discard the warmup draws
+    return draws[warmup:]
+
+
+ar1_data = simulate_ar(10, -0.9, 0)
+
+fig, ax = plt.subplots(figsize=(10, 3))
+ax.set_title("Generated Autoregressive Timeseries", fontsize=15)
+ax.plot(ar1_data);
+```
+
+## Specifying the Model
+
+We'll walk through the model step by step and then generalise the pattern into a function that can be used to take increasingly complex structural combinations of components.
+
+```{code-cell} ipython3
+## Set up a dictionary for the specification of our priors
+## We set up the dictionary to specify size of the AR coefficients in
+## case we want to vary the AR lags.
+priors = {
+    "coefs": {"mu": [10, 0.2], "sigma": [0.1, 0.1], "size": 2},
+    "sigma": 8,
+    "init": {"mu": 9, "sigma": 0.1, "size": 1},
+}
+
+## Initialise the model
+with pm.Model() as AR:
+    pass
+
+## Define the time interval for fitting the data
+t_data = list(range(len(ar1_data)))
+## Add the time interval as a mutable coordinate to the model to allow for future predictions
+AR.add_coord("obs_id", t_data, mutable=True)
+
+with AR:
+    ## Data containers to enable prediction
+    t = pm.MutableData("t", t_data, dims="obs_id")
+    y = pm.MutableData("y", ar1_data, dims="obs_id")
+
+    # The first coefficient will be the constant term but we need to set priors for each coefficient in the AR process
+    coefs = pm.Normal("coefs", priors["coefs"]["mu"], priors["coefs"]["sigma"])
+    sigma = pm.HalfNormal("sigma", priors["sigma"])
+    # We need one init variable for each lag, hence size is variable too
+    init = pm.Normal.dist(
+        priors["init"]["mu"], priors["init"]["sigma"], size=priors["init"]["size"]
+    )
+    # Steps of the AR model minus the lags required
+    ar1 = pm.AR(
+        "ar",
+        coefs,
+        sigma=sigma,
+        init_dist=init,
+        constant=True,
+        steps=t.shape[0] - (priors["coefs"]["size"] - 1),
+    )
```

Comment:
I'd recommend adding the `obs_id` coeffs here for the `ar` and `likelihood` nodes. That way the graphviz below will render them even better in the same plate as `t` and `y`.

Reply:
The structure is a little cleaner now: 
<img width="560" alt="image" src="https://user-images.githubusercontent.com/10588167/200553198-b20442ca-2e7d-4e16-b6cb-16ae9cfb0d3d.png">